src/net_processing.cpp

```diff
@@ -2591,6 +2592,50 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
         }
     }
 
+
+    else if (strCommand == NetMsgType::GETFRAUD)
+    {
+        CBlockLocator locator;
+        vRecv >> locator;
+        CBlockIndex *pindex = NULL;
+        for (const uint256& hash : locator.vHave) {
+            BlockMap::iterator mi = mapBlockIndex.find(hash);
+            if (mi == mapBlockIndex.end()) {
+                continue;
+            }
+            for (pindex = mi->second; pindex; pindex = pindex->pprev) {
+                if (pindex->IsValid(BLOCK_VALID_SCRIPTS)) {
+                    connman.PushMessage(pfrom, msgMaker.Make(NetMsgType::FRAUD, pindex->GetBlockHash(), VARINT(0)));
+                    break;
+                }
+                if (pindex->nStatus & BLOCK_FAILED_VALID) {
+                    if ((pindex->nStatus & BLOCK_VALID_MASK) >= BLOCK_VALID_TRANSACTIONS) {
+                        // These are only reached when block size/weight are known to be good
+                    } else {
+                        CBlock block;
+                        const bool ret = ReadBlockFromDisk(block, pindex, chainparams.GetConsensus());
```

Comment:
probably need to test `(pIndex->nStatus & BLOCK_HAVE_DATA)` in case of pruned node.